### DISCUSSION TITLE
Question about custom path & backup

### DISCUSSION THREAD

**Discussion by piperun** [simolus3/drift#3266] [Q&A] [answered]
Hi!
I'm wondering just how you're suppose to set a custom path (in my case I just want to store the database in production in `getApplicationSupportDirectory()` instead of `getApplicationDocumentsDirectory()`.

Also when it comes to importing/exporting the database I didn't quite get the documentation's example.

**Comment by simolus3** [comment]
> I'm wondering just how you're suppose to set a custom path (in my case I just want to store the database in production in `getApplicationSupportDirectory()` instead of `getApplicationDocumentsDirectory()`.

Are you using `drift_flutter`? In that case, use the [`databasePath`](https://pub.dev/documentation/drift_flutter/latest/drift_flutter/DriftNativeOptions/databasePath.html) argument to `DriftNativeOptions`.

> Also when it comes to importing/exporting the database I didn't quite get the documentation's example.

The example mainly shows how to use existing databases that you might have as an asset for initial data. Are you looking for a way to implement save/restore functionality based on database files?

**Comment by piperun** [comment]
>Are you using drift_flutter? In that case, use the [databasePath](https://pub.dev/documentation/drift_flutter/latest/drift_flutter/DriftNativeOptions/databasePath.html) argument to DriftNativeOptions.

Yes, sorry for asking this but, where exactly can you access `DriftNativeOptions`? I tried looking into `driftDatabase`:
```
  static QueryExecutor _openConnection() {
    // `driftDatabase` from `package:drift_flutter` stores the database in
    // `getApplicationDocumentsDirectory()`.
    return driftDatabase();
  }
```
But could only see: `DriftWebOptions`.

>The example mainly shows how to use existing databases that you might have as an asset for initial data. Are you looking for a way to implement save/restore functionality based on database files?

Correct, essentially I just want to have the ability to import/export said database (and by extension be able to backup the database)..

**Reply by simolus3** [reply]
> Yes, sorry for asking this but, where exactly can you access `DriftNativeOptions`? I tried looking into `driftDatabase`:

It's available [here](https://pub.dev/documentation/drift_flutter/latest/drift_flutter/DriftNativeOptions-class.html), can you check that you're using `drift_flutter` version 0.2.0?

> Correct, essentially I just want to have the ability to import/export said database (and by extension be able to backup the database)..

What's needed to export databases is [described in the first snippet here](https://drift.simonbinder.eu/docs/examples/existing_databases/#exporting-a-database).
To import a database (given as a file perhaps), you'd

1. First close your existing app database.
2. Then, open the database serving as a source to import. You can use the `sqlite3` package for this directly, this step doesn't need drift.
3. Issue a `VACUUM INTO ?` statement with `?` bound to the path of your app database.
4. Open your app database again.

**Reply by piperun** [reply]
Ah, thanks a lot!

Could I use drift to also import said database?

I did find this:
```dart


    void importDatabase() async
    {
    FilePickerResult? result = await FilePicker.platform.pickFiles(
    type: FileType.custom,
    allowedExtensions: ["sqlite"]
    );
    if (result != null) {
    File selectedFile = File(result.files.single.path!);
    print("file path ${selectedFile.path}");
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(p.join(dbFolder.path, getUsername()! + 'crmpharma.sqlite'));
    await file.copy(p.join(dbFolder.path, "backup." + getUsername()! + 'crmpharma.sqlite'));
    var updatedDB = await selectedFile.readAsBytes();
    await file.writeAsBytes(updatedDB);
    await showMessage("Operation success", "Database imported successfully!");
    Navigator.of(context).pushAndRemoveUntil(MaterialPageRoute(builder: (context) =>SetupPage()), (Route route)=>false);
    } else {
    }
    }

//Export

    Future exportDatabaseInto(File file) async {
    await file.parent.create(recursive: true);
    if (file.existsSync()) {
    file.deleteSync();
    }
    await customStatement('VACUUM INTO ?', [file.path]);
    }

```
From: https://github.com/simolus3/drift/issues/2307

**Reply by simolus3** [reply]
Copying over file contents generally works, but a potential issue is that, with [write-ahead logging](https://www.sqlite.org/wal.html) enabled, the database is stored in two files. Forgetting to copy the `-wal` file over can break the database, which is why I recommend using `VACUUM` instead:

```dart
void importDatabase() async {
  FilePickerResult? result = await FilePicker.platform
      .pickFiles(type: FileType.custom, allowedExtensions: ["sqlite"]);
  if (result != null) {
    final selectedPath = result.files.single.path!;
    // Replace this with your actual database path
    final dbFolder = await getApplicationDocumentsDirectory();
    final targetPath =
        p.join(dbFolder.path, getUsername()! + 'crmpharma.sqlite');
    if (await File(targetPath).exists()) {
      await File(targetPath).delete();
    }

    await Isolate.run(() async {
      final database = sqlite3.open(selectedPath);
      database.execute('VACUUM INTO ?', [targetPath]);
      database.dispose();
    });
  } else {}
}

```

**Reply by piperun** [reply]
Hi! Sorry for late reply, I just now got time to read things more thoroughly and test things out, `drift_flutter` was version `0.1.0` which I assume is why I couldn't see DriftNativeOptions.

I toyed around with your code example (and the one I found):
```dart
void importDatabase(String dbPath) async {
  if (dbPath.isNotEmpty) {
    final selectedPath = dbPath;
    // Replace this with your actual database path
    final dbFolder = await getApplicationDocumentsDirectory();
    final targetPath = p.join(dbFolder.path, "test_backup.sqlite");
    if (await File(targetPath).exists()) {
      await File(targetPath).delete();
    }

    await Isolate.run(() async {
      final database = sqlite3.open(selectedPath);
      database.execute("VACUUM INTO ?", [targetPath]);
      database.dispose();
    });
  } else {}
}
```
Doesn't this work more as an export function? since we create a new file `test_backup.sqlite`?
I presume then that the actual import function would just be something like `database.open("path")`?

**Reply by simolus3** [reply]
Sorry for being unclear, I kind of copied stuff from the earlier responses around. `test_backup.sqlite` should be the path you're using for your regular in-app database.

To import a database from `importPath` to `appDatabasePath`, the steps are:

1. Close your app database; delete file at `appDatabasePath`.
2. Use `package:sqlite3` to open the database at `importPath`
3. Run the `VACUUM INTO appDatabasePath` on the database opened in step 2. This re-creates a compact copy of the database you want to import at the correct location in your app.
4. Then, close the database from step 2 and re-open your normal app database at `appDatabasePath`.

Ideally you'll want to run steps 2 and 3 in a background isolate because `package:sqlite3` does synchronous IO.

**Reply by piperun** [reply]
No problem!

Forgive my ignorance here, but I realized just now that theoretically I could do:
`AppDatabase(name: "import_db_file", path: customPath)` and close when needed to, that would count as an import right?

But I assume there has to be some reason (and ideally I would like if there was a small section in the documentation about importing/exporting the database) to choose:
```dart
Future<void> importDatabase(String importPath) async {
  if (importPath.isNotEmpty) {
    final dbFolder = await getApplicationDocumentsDirectory();
    final targetPath = p.join(dbFolder.path, "test_backup.sqlite");
    if (await File(targetPath).exists()) {
      await File(targetPath).delete();
    }

    await Isolate.run(() async {
      final database = sqlite3.open(importPath);
      database.execute("VACUUM INTO ?", [targetPath]);
      database.dispose();
    });
  } else {}
}
// other file:
AppDatabase(name: "test_backup");
```
Over something like:
(Wrote this mostly to ensure it's simple enough to get the main idea across that we can swap out the database, in the real application we would most likely have a Provider class making this possible & have proper "default path string" in a setting)
```dart
Future<void> openDefaultDatabase() async {
  final database = AppDatabase(name: "default_db");
  await database.doTasks();
  await database.close();
}

Future<void> openAnyDatabase(String name, Future<String> Function ()? customPath) async {
    final database = AppDatabase(name: name, DriftNativeOptions(
          databasePath: customPath,
        ));
  await database.doTasks();
  await database.close();
}
```
Or have I completely misunderstood this?

**Reply by simolus3** [reply]
> `AppDatabase(name: "import_db_file", path: customPath)` and close when needed to, that would count as an import right?

Yes, but then you need to remember the imported path if you want the import to survive app restarts. Depending on the OS, you might also not have write permissions for the selected file (e.g. a read-only copy shared with your app).
That's the only reason for copy contents over, if that's not relevant for you, just opening the imported file (perhaps as read-only) directly is indeed easier and no other downsides come to mind.

**Reply by piperun** [reply]
Thank you very much for this information, my last question would then be, if we then follow the documentation's recommendation, is this why drift defaults to `getApplicationDocumentsDirectory()` instead of `  getApplicationSupportDirectory()`? To essentially ensure we always will have write access?

**Reply by simolus3** [reply, marked as answer]
`getApplicationSupportDirectory()` would have arguably been a better choice. There is no strong reason for this outside of backwards compatibility now, you can safely use a custom path in the support directory.
